enterprise/reporting/error_reporting.go

```diff
@@ -317,82 +339,109 @@ func (edr *ErrorDetailReporter) mainLoop(ctx context.Context, c types.SyncerConf
 	edr.errorDetailReportsQueryTime = edr.stats.NewTaggedStat("error_detail_reports_query_time", stats.TimerType, tags)
 	edr.edReportingRequestLatency = edr.stats.NewTaggedStat("error_detail_reporting_request_latency", stats.TimerType, tags)
 
-	// In infinite loop
-	// Get Reports
-	// Aggregate
-	// Send in a separate go-routine
-	// Delete in a separate go-routine
-	for {
-		if ctx.Err() != nil {
-			edr.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-			return
-		}
-		requestChan := make(chan struct{}, edr.maxConcurrentRequests.Load())
-		loopStart := time.Now()
-		currentMs := time.Now().UTC().Unix() / 60
+	var lastReportedAtTime atomic.Time
+	lastReportedAtTime.Store(time.Now())
 
-		getReportsStart := time.Now()
-		reports, reportedAt := edr.getReports(ctx, currentMs, c.ConnInfo)
-		getReportsTimer.Since(getReportsStart)
-		getReportsSize.Observe(float64(len(reports)))
+	g, ctx := errgroup.WithContext(ctx)
 
-		if len(reports) == 0 {
-			select {
-			case <-ctx.Done():
+	g.Go(func() error {
+		return edr.emitLagMetric(ctx, c, &lastReportedAtTime)
+	})
+
+	g.Go(func() error {
+		// In infinite loop
+		// Get Reports
+		// Aggregate
+		// Send in a separate go-routine
+		// Delete in a separate go-routine
+		for {
+			if ctx.Err() != nil {
 				edr.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
-				return
-			case <-time.After(edr.sleepInterval.Load()):
+				return ctx.Err()
 			}
-			continue
-		}
-
-		aggregationStart := time.Now()
-		metrics := edr.aggregate(reports)
-		aggregateTimer.Since(aggregationStart)
-		getAggregatedReportsSize.Observe(float64(len(metrics)))
-
-		errGroup, errCtx := errgroup.WithContext(ctx)
-		for _, metric := range metrics {
-			metricToSend := metric
-			requestChan <- struct{}{}
-			if errCtx.Err() != nil {
-				// if any of errGroup's goroutines fail - don't send anymore requests for this batch
-				break
+			requestChan := make(chan struct{}, edr.maxConcurrentRequests.Load())
+			loopStart := time.Now()
+			currentMs := time.Now().UTC().Unix() / 60
+
+			getReportsStart := time.Now()
+			reports, reportedAt := edr.getReports(ctx, currentMs, c.ConnInfo)
+			if ctx.Err() != nil {
+				edr.log.Errorw("getting reports", "error", ctx.Err())
+				select {
+				case <-ctx.Done():
+					edr.log.Infof("stopping mainLoop for syncer %s : %s", c.Label, ctx.Err())
+					return ctx.Err()
+				case <-time.After(edr.mainLoopSleepInterval.Load()):
+				}
+				continue
 			}
-			errGroup.Go(func() error {
-				err := edr.sendMetric(errCtx, c.Label, metricToSend)
-				if err != nil {
-					edr.log.Error("Error while sending to Reporting service:", err)
+			lastReportedAtTime.Store(time.Unix(reportedAt*60, 0))
```

Comment:
let's move this after checking for `reports` length to be zero (after line 391).